### PR TITLE
distributed regression implementation + unit tests

### PR DESCRIPTION
**Pull request by rishic3**
- Changes:
   * Implemented distributed make_regression
   * Implemented unit testing for both default case (X is generated from standard normal) and low rank case (X is generated from low_rank_matrix)

### CODE REVIEW

**Comment by rishic3** [comment]
Regression runtime test:
1,000,000 rows, 3000 columns, float32, 50 partitions, 20g driver memory, 16 threads
-> ~9 minutes 5 seconds

**Comment by rishic3** [comment]
build

File: python/benchmark/test_gen_data.py [hunk 1]
```diff
@@ -103,3 +103,91 @@ def test_make_low_rank_matrix(dtype: str) -> None:
...
+    # Effective rank = None
+    input_args = [
+        "--num_rows",
+        "100",
+        "--num_cols",
+        "10",
+        "--dtype",
+        dtype,
+        "--output_dir",
+        "temp",
```
Comment (wbo4958) [inline python/benchmark/test_gen_data.py:164]: we have a fixture tmp_path that you can use directly. please refer to https://github.com/NVIDIA/spark-rapids-ml/blob/branch-23.06/python/tests/conftest.py#L54-L58

File: python/benchmark/test_gen_data.py [hunk 2]
```diff
@@ -103,3 +103,91 @@ def test_make_low_rank_matrix(dtype: str) -> None:
...
+    # Effective rank = 5
+    input_args = [
+        "--num_rows",
+        "100",
+        "--num_cols",
+        "10",
+        "--dtype",
+        dtype,
+        "--output_dir",
+        "temp",
```
Comment (wbo4958) [inline python/benchmark/test_gen_data.py:119]: please use `tmp_path` directly.

File: python/benchmark/test_gen_data.py [hunk 3]
```diff
@@ -103,3 +103,91 @@ def test_make_low_rank_matrix(dtype: str) -> None:
 
         _, s, _ = svd(X)
         assert sum(s) - 5 < 0.1, "X rank is not approximately 5"
+
+
+@pytest.mark.parametrize("dtype", ["float32", "float64"])
+def test_make_regression_low_rank(dtype: str) -> None:
+    # Effective rank = 5
```
Comment (wbo4958) [inline python/benchmark/test_gen_data.py:110]: Seems you can combine test_make_regression_low_rank and test_make_regression_well_conditioned with effictive_rank parameterized.

File: python/benchmark/gen_data_distributed.py [hunk 4]
```diff
@@ -232,6 +234,152 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+            # Generate a low-rank, fat tail input set.
+            X, _ = LowRankMatrixDataGen(lrm_input_args).gen_dataframe(spark)
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from LowRankMatrix;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+        else:
+            # Randomly generate a well-conditioned input set.
+            X = spark.createDataFrame(
+                RandomRDDs.normalVectorRDD(
```
Comment (wbo4958) [inline python/benchmark/gen_data_distributed.py:319]: can it generate the regression datasets?

File: python/benchmark/test_gen_data.py (hunk 1 above)
Comment (rishic3) [reply python/benchmark/test_gen_data.py:164]: The path isn't actually being used (only provided since it's required) as the test doesn't execute main, just checks the dataframe directly. Should I still bother using the fixture?

File: python/benchmark/test_gen_data.py (hunk 3 above)
Comment (rishic3) [reply python/benchmark/test_gen_data.py:110]: To invoke the well-conditioned test, I need to leave off "--effective_rank" entirely; the parser expects either an int or no arg at all. I could combine them into one test, but that would still require defining two sets of input_args and having an if statement.

File: python/benchmark/gen_data_distributed.py (hunk 4 above)
Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:319]: Both generate the regression set; the effective_rank just determines how biased the nput set distribution is.

**Comment by rishic3** [comment]
build

File: python/benchmark/gen_data_distributed.py [hunk 5]
```diff
@@ -232,6 +234,152 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+        noise = params.get("noise", 0.0)
+        shuffle = params.get("shuffle", True)
+        effective_rank = params.get("effective_rank", None)
+        n_informative = params.get("n_informative", 10)
+        n_targets = params.get("n_targets", 1)
+
+        # Input set is either well conditioned (default) or has a low rank fat tail singular profile (see LowRankMatrixDataGen).
+        # Output is generated by applying a (potentially biased) random linear regression model to the input, with n_informative
+        # nonzero regressors and some gaussian centered noise with adjustable scale.
+        # Code adapted from sklearn.datasets.make_regression().
```
Comment (lijinf2) [inline python/benchmark/gen_data_distributed.py:289]: Need to specify "Comments (line 286-288) come from sklearn.datasets.make_regression()." as well.

File: python/benchmark/gen_data_distributed.py [hunk 6]
```diff
@@ -232,6 +234,152 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+
+        # Generate ground truth upfront.
+        ground_truth = np.zeros((cols, n_targets))
+        ground_truth[:n_informative, :] = 100 * generator.uniform(
+            size=(n_informative, n_targets)
+        )
+
+        if shuffle:
```
Comment (lijinf2) [inline python/benchmark/gen_data_distributed.py:340]: Why shuffle is required in this class?

File: python/benchmark/gen_data_distributed.py (hunk 5 above)
Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:289]: Not sure what you mean - as in, I should specify which exact code lines were adapted from sklearn?

File: python/benchmark/gen_data_distributed.py (hunk 6 above)
Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:343]: Low rank matrix / randomRDDs don't have any shuffling of their own, so row-wise shuffle per partition is meant to reduce some of the potential sample generation bias. Column-wise shuffle is to prevent same indices to be set as informative features every time.

**Comment by lijinf2** [comment]
LGTM

File: python/benchmark/test_gen_data.py (hunk 3 above)
Comment (eordentlich) [reply python/benchmark/test_gen_data.py:110]: You can conditionally construct the args list from multiple pieces depending on the value of effective_rank parameter in a parameterized test.

**Comment by rishic3** [comment]
build

File: python/benchmark/gen_data_distributed.py [hunk 7]
```diff
@@ -232,6 +234,155 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+            ]
+            # Generate a low-rank, fat tail input set.
+            X, _ = LowRankMatrixDataGen(lrm_input_args).gen_dataframe(spark)
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from LowRankMatrix;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+        else:
+            # Randomly generate a well-conditioned input set.
+            X = spark.createDataFrame(
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:321]: Will this be at a different variance than X from low_rank_matrix?

File: python/benchmark/gen_data_distributed.py [hunk 8]
```diff
@@ -232,6 +234,155 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+                X_p = pdf.to_numpy()
+                n_partition_rows = X_p.shape[0]
+
+                if shuffle:
+                    # Column-wise shuffle (global)
+                    X_p[:, :] = X_p[:, indices]
+
+                y = np.dot(X_p, ground_truth) + bias
+                if noise > 0.0:
+                    y += generator.normal(scale=noise, size=y.shape)
```
Comment (eordentlich) [inline python/benchmark/gen_data_distributed.py:361]: So due to the unit column normalization of low rank matrix the 'signal to noise' ratio will depend on the number of examples for a given 'noise' parameter.  Is that how sklearn works?

Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:361]: Not sure what you mean here by the number of examples for a given noise parameter.

File: python/benchmark/gen_data_distributed.py (hunk 7 above)
Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:321]: I believe it should be, since the variance from an X in low_rank_matrix will depend on the magnitudes of the singular values, which will vary by the degree of tail_strength.

File: python/benchmark/gen_data_distributed.py (hunk 8 above)
Comment (eordentlich) [reply python/benchmark/gen_data_distributed.py:361]: Will the variance of the predict value decrease with the number examples generated (e.g. due the orthonormality constraint on the output columns in low-rank-matrix)?

File: python/benchmark/gen_data_distributed.py (hunk 7 above)
Comment (eordentlich) [reply python/benchmark/gen_data_distributed.py:321]: Sorry, I meant in terms of how the variance scales/depends on the number of examples generated.

File: python/benchmark/gen_data_distributed.py (hunk 8 above)
Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:361]: Yes, I figure it would. SKLearn's implementation allows this to happen, and they don't ensure any norm on the default cause (with generator.standard_normal) to make the two methods norm-consistent. But in both cases the relative variances of X and y will scale the same, so I imagine it would be fine?

Comment (rishic3) [reply python/benchmark/gen_data_distributed.py:361]: And yes, this version is consistent with what sklearn is doing. Normed X with low_rank_matrix, and no norm via the standard normal.

**Comment by eordentlich** [comment]
build

**Comment by eordentlich** [comment]
ci failing formatting

**Comment by rishic3** [comment]
build

**Comment by rishic3** [comment]
build

**Comment by rishic3** [comment]
build

File: python/benchmark/test_gen_data.py [hunk 9]
```diff
@@ -103,3 +103,50 @@ def test_make_low_rank_matrix(dtype: str) -> None:
 
         _, s, _ = svd(X)
         assert sum(s) - 5 < 0.1, "X rank is not approximately 5"
+
+
+@pytest.mark.parametrize(
```
Comment (eordentlich) [inline python/benchmark/test_gen_data.py:108]: Missed this before.  Check other examples in the repo on how to set up a multi-parametrized test using multiple parametrize statements.  Don't need to enumerate the combinations as it is done automatically.

**Comment by rishic3** [comment]
build

**Review by eordentlich — APPROVED** [review]
👍
